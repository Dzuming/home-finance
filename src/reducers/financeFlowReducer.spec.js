import reducer from './financeFlowReducer';
import deepFreeze from 'deep-freeze';
import {
  setSpending, setProfit, addSpending, removeSpending, editSpending,
  addProfit, removeProfit
} from '../actions/financeFlowActions';

const initialState = deepFreeze(reducer({spending: [], profit: []}, {type: 'INIT'}));
const spending = {
  id: 1,
  description: 'Zakupy',
  value: 300,
  user_id: 1,
  created_at: '2018-01-01 12:12:12',
  category: {id: 1, name: 'jedzenie'}
};
const profit = {
  id: 1,
  description: 'Zakupy',
  value: 300,
  user_id: 1,
  created_at: '2018-01-01 12:12:12',
  category: {id: 1, name: 'jedzenie'}
};
const message = 'test';
describe('spending reducer', () => {
  it('should handle unknown actions', () => {
    expect(reducer(initialState, {type: 'FAKE'})).toBe(initialState);
  });

  describe('get action', () => {
    it('should get spending', () => {
      expect(reducer(initialState, setSpending([spending]))).toMatchSnapshot();
    });
  });

  describe('add action', () => {
    it('should add spending to empty list', () => {
      expect(reducer(initialState, addSpending({spending, message}))).toMatchSnapshot();
    });

    it('should add spending to a non-empty list', () => {
      const nonEmptyState = deepFreeze(reducer(initialState, addSpending({spending, message})));

      expect(reducer(nonEmptyState, addSpending({spending, message}))).toMatchSnapshot();
    });
  });

  describe('delete action', () => {
    const baseStateSpending = deepFreeze(
      [{spending: {...spending, id: 1}}, {spending: {...spending, id: 2}}, {spending: {...spending, id: 3}}]
        .reduce((state, spending) => reducer(state, addSpending(spending)), initialState)
    );

    it('should delete spending when exists', () => {
      expect(reducer(baseStateSpending, removeSpending({id: 1}))).toMatchSnapshot();
    });

    it('should not delete spending when it doesn\'t exist', () => {
      expect(reducer(baseStateSpending, removeSpending({id: 4}))).toMatchSnapshot();
    });
  });

  describe('edit action', () => {
    const baseState = deepFreeze(
      [{spending: {...spending, id: 1}}, {spending: {...spending, id: 2}}, {spending: {...spending, id: 3}}]
        .reduce((state, spending) => reducer(state, addSpending(spending)), initialState)
    );
    const user = {
      id: 1,
      name: 'test',
    };
    localStorage.setItem('user', JSON.stringify(user));
    it('should edit spending when exists', () => {
      expect(reducer(baseState, editSpending('test', 1, {data: {description: 'OpÅ‚ata rachunki'}}))).toMatchSnapshot();
    });
  });
});

describe('profit reducer', () => {
  it('should handle unknown actions', () => {
    expect(reducer(initialState, {type: 'FAKE'})).toBe(initialState);
  });

  describe('get action', () => {
    it('should get profit', () => {
      expect(reducer(initialState, setProfit([profit]))).toMatchSnapshot();
    });
  });

  describe('add action', () => {
    it('should add profit to empty list', () => {
      expect(reducer(initialState, addProfit({profit, message}))).toMatchSnapshot();
    });

    it('should add profit to a non-empty list', () => {
      const nonEmptyState = deepFreeze(reducer(initialState, addProfit({profit, message})));

      expect(reducer(nonEmptyState, addProfit({profit, message}))).toMatchSnapshot();
    });
  });

  describe('delete action', () => {
    const baseStateProfit = deepFreeze(
      [{profit: {...profit, id: 1}}, {profit: {...profit, id: 2}}, {profit: {...profit, id: 3}}]
        .reduce((state, profit) => reducer(state, addProfit(profit)), initialState)
    );
    it('should delete profit when exists', () => {
      expect(reducer(baseStateProfit, removeProfit({id: 1}))).toMatchSnapshot();
    });

    it('should not delete profit when it doesn\'t exist', () => {
      expect(reducer(baseStateProfit, removeProfit({id: 4}))).toMatchSnapshot();
    });
  });

  describe('edit action', () => {
  });
});
