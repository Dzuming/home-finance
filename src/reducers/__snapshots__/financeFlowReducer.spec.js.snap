// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profit reducer add action should add profit to a non-empty list 1`] = `
Object {
  "message": "test",
  "profit": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
  ],
  "spending": Array [],
}
`;

exports[`profit reducer add action should add profit to empty list 1`] = `
Object {
  "message": "test",
  "profit": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
  ],
  "spending": Array [],
}
`;

exports[`profit reducer delete action should delete profit when exists 1`] = `
Object {
  "message": Object {
    "id": 1,
  },
  "profit": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "profit": 300,
    },
  ],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "profit": 300,
    },
  ],
}
`;

exports[`profit reducer delete action should not delete profit when it doesn't exist 1`] = `
Object {
  "message": Object {
    "id": 4,
  },
  "profit": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "profit": 300,
    },
  ],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "profit": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "profit": 300,
    },
  ],
}
`;

exports[`profit reducer get action should get profit 1`] = `
Object {
  "profit": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "profit": 300,
    },
  ],
  "spending": Array [],
}
`;

exports[`spending reducer add action should add spending to a non-empty list 1`] = `
Object {
  "message": "test",
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
  ],
}
`;

exports[`spending reducer add action should add spending to empty list 1`] = `
Object {
  "message": "test",
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
  ],
}
`;

exports[`spending reducer delete action should delete spending when exists 1`] = `
Object {
  "message": Object {
    "id": 1,
  },
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "spending": 300,
    },
  ],
}
`;

exports[`spending reducer delete action should not delete spending when it doesn't exist 1`] = `
Object {
  "message": Object {
    "id": 4,
  },
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "spending": 300,
    },
  ],
}
`;

exports[`spending reducer edit action should edit spending when exists 1`] = `
Object {
  "message": "test",
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Op≈Çata rachunki",
      "id": 1,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 2,
      "spending": 300,
    },
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 3,
      "spending": 300,
    },
  ],
}
`;

exports[`spending reducer get action should get spending 1`] = `
Object {
  "profit": Array [],
  "spending": Array [
    Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": "Zakupy",
      "id": 1,
      "spending": 300,
    },
  ],
}
`;
